src/sqlfluff/rules/L003.py

```diff
@@ -56,241 +198,169 @@ def _make_indent(
+                current_line = cls._process_line_indents(
+                    elem, current_line, tab_space_size
+                )
+                continue
+
+            elif elem.is_code and current_line.hanger_pos is None:
+                current_line.hanger_pos = cls._indent_size(
+                    current_line.line_buffer[:-1], tab_space_size=tab_space_size
+                )
 
         # If we get to the end, and still have a buffer, add it on
-        if line_buffer:
-            result_buffer[line_no] = {
-                "line_no": line_no,
-                "templated_line": templated_line,
-                "line_buffer": line_buffer,
-                "indent_buffer": indent_buffer,
-                "indent_size": indent_size,
-                "indent_balance": this_indent_balance,
-                "hanging_indent": line_indent_stack.pop()
-                if line_indent_stack
-                else None,
-                "clean_indent": clean_indent,
-            }
+        if current_line.line_buffer:
+            result_buffer[line_no] = current_line.from_current_line(
+                line_no,
+                templated_file,
+            )
         return result_buffer
 
+    @classmethod
+    def _process_line_indents(
+        cls,
+        elem: BaseSegment,
+        current_line: _LineSummary,
+        tab_space_size: int,
+    ) -> _LineSummary:
+        current_line.indent_balance += elem.indent_val  # type: ignore
+        if elem.indent_val > 0:  # type: ignore
+            # Keep track of the indent at the last ... indent
+            current_line.line_indent_stack.append(
+                cls._indent_size(
+                    current_line.line_buffer, tab_space_size=tab_space_size
+                )
+            )
+            current_line.hanger_pos = None
+            return current_line
+        # this is a dedent, we could still have a hanging indent,
+        # but only if there's enough on the stack
+        if current_line.line_indent_stack:
+            current_line.line_indent_stack.pop()
+        return current_line
+
+    @classmethod
+    def _process_pre_anchor(
+        cls,
+        elem: BaseSegment,
+        current_line: _LineSummary,
+        tab_space_size: int,
+    ) -> _LineSummary:
+        if elem.is_whitespace:
+            current_line.indent_buffer.append(elem)
+            return current_line
+        if elem.is_meta and elem.indent_val != 0:  # type: ignore
+            current_line.indent_balance += elem.indent_val  # type: ignore
+            if elem.indent_val > 0:  # type: ignore
+                # a "clean" indent is one where it contains
+                # an increase in indentation? Can't quite
+                # remember the logic here. Let's go with that.
+                current_line.clean_indent = True
+            return current_line
+
+        current_line.set_state_as_of_anchor(elem, tab_space_size)
+        return current_line
```

Comment:
Same concern as noted elsewhere -- to avoid confusion, maybe don't return the same object we were passed, then consider adding back the `return` once it's returning a copy.